hack/generator/pkg/astmodel/types.go

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import "fmt"
+
+// Types is the set of all types being generated
+type Types map[TypeName]TypeDefinition
+
+// Add adds a type to the set, with safety check that it has not already been defined
+func (types Types) Add(def TypeDefinition) {
+	key := def.Name()
+	if _, ok := types[key]; ok {
+		panic(fmt.Sprintf("type already defined: %v", key))
+	}
+
+	types[key] = def
+}
+
+// AddAll adds multiple types to the set, with the same safety check as Add() to panic if a duplicate is included
+func (types Types) AddAll(otherTypes Types) {
+	for _, t := range otherTypes {
+		types.Add(t)
+	}
+}
+
+// Where returns a new set of types including only those that satisfy the predicate
+func (types Types) Where(predicate func(definition TypeDefinition) bool) Types {
+	result := make(Types, len(types))
+	for _, t := range types {
+		if predicate(t) {
+			result[t.Name()] = t
+		}
+	}
+
+	return result
+}
+
+// Except returns a new set of types including only those not present in otherTypes
+func (types Types) Except(otherTypes Types) Types {
+	return types.Where(func(def TypeDefinition) bool {
+		return !otherTypes.Contains(def)
+	})
+}
+
+// Contains returns true if the set contains the passed definition
```

Comment:
comment is untrue, it checks if the set contains something with the same _name_ 🙂

Reply:
Changed to accept a name, and updated the comment.